runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java

```diff
@@ -561,102 +623,148 @@ public void onMerge(Collection<W> toBeMerged, W mergeResult) throws Exception {
       // cannot take a trigger state from firing to non-firing.
       // (We don't actually assert this since it is too slow.)
     }
-
-    return triggerableWindows;
   }
 
-  /**
-   * Called when an end-of-window, garbage collection, or trigger-specific timer fires.
-   */
-  public void onTimer(TimerData timer) throws Exception {
-    // Which window is the timer for?
-    checkArgument(timer.getNamespace() instanceof WindowNamespace,
-        "Expected timer to be in WindowNamespace, but was in %s", timer.getNamespace());
-    @SuppressWarnings("unchecked")
-    WindowNamespace<W> windowNamespace = (WindowNamespace<W>) timer.getNamespace();
-    W window = windowNamespace.getWindow();
-    ReduceFn<K, InputT, OutputT, W>.Context directContext =
-        contextFactory.base(window, StateStyle.DIRECT);
-    ReduceFn<K, InputT, OutputT, W>.Context renamedContext =
-        contextFactory.base(window, StateStyle.RENAMED);
+  private class TimerContext {
```

Comment:
Can you document this a bit more? We have a lot of `XYZContext` classes floating around, and they don't all have an obvious meaning.

This actually looks it could be an `EnrichedTimer` as it basically adds some queries that you can ask about a particular `TimerData`.